daemon/containerd/image_pull.go

```diff
@@ -48,6 +50,34 @@ func (i *ImageService) PullImage(ctx context.Context, image, tagOrDigest string,
 	resolver, _ := i.newResolverFromAuthConfig(ctx, authConfig)
 	opts = append(opts, containerd.WithResolver(resolver))
 
+	skipExistingManifests := images.HandlerFunc(func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		if desc.MediaType == images.MediaTypeDockerSchema2Manifest || desc.MediaType == ocispec.MediaTypeImageManifest {
+			if _, err := i.client.ContentStore().Info(ctx, desc.Digest); err == nil {
+				return nil, images.ErrSkipDesc
+			}
+		}
+		return nil, nil
+	})
```

Comment:
>  // ErrSkipDesc is used to skip processing of a descriptor and
>	// its descendants.
>	ErrSkipDesc = errors.New("skip descriptor")


Won't this prevent pull from fetching the descendants of manifest if they're missing?

For example, if we already have a manifest for other platform, but don't have the blobs, then we won't be able to pull anything under this manifest (blobs, config), because the whole manifest and its children will be skipped.

Reply:
It will yes, the idea is that if you have the manifest, then you must have the layers too, if you pull for another platform you don't have the manifest yet.

Reply:
We do have one issue though: if we start a pull and then cancel it we have the manifest in the store but not the layers, so `i.client.Pull` fails to unpack. We'll have to fix that

Reply:
> the idea is that if you have the manifest, then you must have the layers too

I don't think this is a safe assumption. There were multiple cases where a manifest was present but we didn't have the config. For example recently I had [this PR](https://github.com/moby/moby/pull/46238) to fix `docker run --platform linux/arm64 busybox` erroring out because somehow I had a manifest for some other platform, but didn't have any other of its blobs.

Reply:
Right but in that linked pr you are removing the content by hand…

I mean, it _should_ be a safe assumption. Guess we’ll have to check that all the required contents are present then. Let me do that 👍

Reply:
Because I didn't know exactly what causes this and how to reproduce this state. And "Use DD with containerd enabled for some time and try to hit this issue" is much worse reproduction step for testing PR than "delete this content by hand" 😄

Reply:
Okay switched it to not use skip